### ISSUE TITLE
Cilium CEP CRD update may cause k8s apiserver to crash

### ISSUE THREAD

**Issue by tgraf** [cilium/cilium#5085]
Cilium: latest/master
Kubernetes: v1.9.8

```
E0802 21:31:39.480184       1 status.go:64] apiserver received an error that is not an metav1.Status: stream error: stream ID 98061; NO_ERROR
I0802 21:31:39.480403       1 trace.go:76] Trace[418895602]: "Update /apis/cilium.io/v2/namespaces/default/ciliumendpoints/echo-7d9f9564df-tjf5z" (started: 2018-08-02 21:30:39.47989544 +0000 UTC m=+200394.485139796) (total time: 1m0.000482196s):
Trace[418895602]: [1m0.000482196s] [1m0.000482196s] END
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0xf2dcf2]

goroutine 30886076 [running]:
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/request.AuditEventFrom(0x0, 0x0, 0xc4255fc740)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/request/context.go:158 +0x22
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.WriteObjectNegotiated(0x0, 0x0, 0x8e5eea0, 0xc4255fc740, 0xc42532ea70, 0x9, 0xc42532ea7a, 0x2, 0x7f4cd07b81b0, 0xc423bc0f00, ...)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:118 +0x1b6
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.ErrorNegotiated(0x0, 0x0, 0x8e3ad60, 0xc4258ad460, 0x8e5eea0, 0xc4255fc740, 0xc42532ea70, 0x9, 0xc42532ea7a, 0x2, ...)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:142 +0x163
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.(*RequestScope).err(0xc4249d4f00, 0x8e3ad60, 0xc4258ad460, 0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:67 +0x10c
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.UpdateResource.func1(0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/update.go:55 +0x1acd
k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/pkg/apiserver.(*crdHandler).ServeHTTP(0xc420448b00, 0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:211 +0xcd1
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*pathHandler).ServeHTTP(0xc422a09e00, 0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:248 +0x26d
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*PathRecorderMux).ServeHTTP(0xc42013be30, 0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:234 +0xa1
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.director.ServeHTTP(0x3ae4107, 0x17, 0xc42082c870, 0xc42013be30, 0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/handler.go:161 +0x6ad
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.(*director).ServeHTTP(0xc420969b00, 0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	<autogenerated>:1 +0x75
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*pathHandler).ServeHTTP(0xc4264d5000, 0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:254 +0x407
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*PathRecorderMux).ServeHTTP(0xc42063e2a0, 0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:234 +0xa1
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.director.ServeHTTP(0x3ac5cb4, 0xe, 0xc42082ccf0, 0xc42063e2a0, 0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/handler.go:161 +0x6ad
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.(*director).ServeHTTP(0xc4208a97e0, 0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	<autogenerated>:1 +0x75
k8s.io/kubernetes/vendor/k8s.io/kube-aggregator/pkg/apiserver.(*proxyHandler).ServeHTTP(0xc424860180, 0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:91 +0x15f
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*pathHandler).ServeHTTP(0xc428d8ab40, 0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:248 +0x26d
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*PathRecorderMux).ServeHTTP(0xc421985570, 0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:234 +0xa1
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.director.ServeHTTP(0x3ac89a1, 0xf, 0xc421a5d9e0, 0xc421985570, 0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/handler.go:161 +0x6ad
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.(*director).ServeHTTP(0xc423abf580, 0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	<autogenerated>:1 +0x75
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthorization.func1(0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/authorization.go:51 +0x37d
net/http.HandlerFunc.ServeHTTP(0xc4229ea320, 0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/usr/local/go/src/net/http/server.go:1918 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithMaxInFlightLimit.func1(0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/maxinflight.go:165 +0x42a
net/http.HandlerFunc.ServeHTTP(0xc422a32f00, 0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/usr/local/go/src/net/http/server.go:1918 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithImpersonation.func1(0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/impersonation.go:49 +0x203a
net/http.HandlerFunc.ServeHTTP(0xc4229ea370, 0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/usr/local/go/src/net/http/server.go:1918 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthentication.func1(0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/authentication.go:79 +0x2b1
net/http.HandlerFunc.ServeHTTP(0xc4229ea3c0, 0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/usr/local/go/src/net/http/server.go:1918 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/request.WithRequestContext.func1(0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/request/requestcontext.go:110 +0xcb
net/http.HandlerFunc.ServeHTTP(0xc423abf5a0, 0x7f4cd07b81b0, 0xc423bc0f00, 0xc425f33200)
	/usr/local/go/src/net/http/server.go:1918 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1(0xc423abf640, 0x8e60ae0, 0xc423bc0f00, 0xc425f33200, 0xc42b5e80c0)
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:93 +0x8d
created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP
	/workspace/anago-v1.9.8-beta.0.40+c138b851781560/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:92 +0x1ab
```

**Comment by joestringer** [comment]
Hit also on k8s 1.7:

[0a9ca6ca_K8sChaosTest_Endpoint_can_still_connect_while_Cilium_is_not_running.zip](https://github.com/cilium/cilium/files/2271440/0a9ca6ca_K8sChaosTest_Endpoint_can_still_connect_while_Cilium_is_not_running.zip)

**Comment by aanm** [comment]
@tgraf after looking at the k8s code that has this panic me and @raybejjani came to a conclusion this is more likely to be a corruption made by Cilium which is going to be fixed in #5118

2nd EDIT: ~It looks like joe saw this bug on #5118, so back to square 0~

3rd EDIT: I noticed on @joestringer PR, the corruption can still happen since the upgrade test happen before the `K8sChaosTest`. By running the old code we could still have some corruption triggering the kube-apiserver to fail. Let's hold #5100 until the #5118 is merged and if this issue will still happen afterwards.

**Comment by aanm** [comment]
@tgraf This issue doesn't seem to be Cilium related and it's fixed since k8s >= 1.11 https://github.com/kubernetes/kubernetes/issues/67245
